libraries/AP_ExternalAHRS/AP_ExternalAHRS.cpp

```diff
@@ -0,0 +1,560 @@
+                       "QffffffLLffffff",
+                       AP_HAL::micros64(),
+                       pkt1.ypr[2], pkt1.ypr[1], pkt1.ypr[0],
+                       pkt1.velNED[0], pkt1.velNED[1], pkt1.velNED[2],
+                       int32_t(pkt1.positionLLA[0]*1.0e7), int32_t(pkt1.positionLLA[1]*1.0e7),
+                       float(pkt1.positionLLA[2]),
+                       pkt1.posU, pkt1.velU,
+                       pkt1.yprU[2], pkt1.yprU[1], pkt1.yprU[0]);
+}
+
+/*
+  process packet type 2
+ */
+void AP_ExternalAHRS::process_packet2(const uint8_t *b)
+{
+    const struct VN_packet2 &pkt2 = *(struct VN_packet2 *)b;
+    const struct VN_packet1 &pkt1 = *last_pkt1;
+
+    last_pkt2_ms = AP_HAL::millis();
+    *last_pkt2 = pkt2;
+
+    gps_data_message_t gps;
+
+    // get ToW in milliseconds
+    gps.gps_week = pkt2.timeGPS / (AP_MSEC_PER_WEEK * 1000000ULL);
+    gps.ms_tow = (pkt2.timeGPS / 1000000ULL) % (60*60*24*7*1000ULL);
+    gps.fix_type = pkt2.GPS1Fix;
+    gps.satellites_in_view = pkt2.numGPS1Sats;
+
+    gps.horizontal_pos_accuracy = pkt1.posU;
+    gps.vertical_pos_accuracy = pkt1.posU;
+    gps.horizontal_vel_accuracy = pkt1.velU;
+
+    gps.hdop = pkt2.GPS1DOP[4];
+    gps.vdop = pkt2.GPS1DOP[3];
+
+    gps.latitude = pkt2.GPS1posLLA[0] * 1.0e7;
+    gps.longitude = pkt2.GPS1posLLA[1] * 1.0e7;
+    gps.msl_altitude = pkt2.GPS1posLLA[2] * 1.0e2;
+
+    gps.ned_vel_north = pkt2.GPS1velNED[0];
+    gps.ned_vel_east = pkt2.GPS1velNED[1];
+    gps.ned_vel_down = pkt2.GPS1velNED[2];
+
+    if (gps.fix_type >= 3 && !origin_set) {
+        origin.lat = gps.latitude;
+        origin.lng = gps.longitude;
+        origin.alt = gps.msl_altitude;
+        origin_set = true;
+    }
+
+    AP::gps().handle_external(gps);
+}
+
+// get serial port number for the uart
+int8_t AP_ExternalAHRS::get_port(void) const
+{
+    if (!uart) {
+        return -1;
+    }
+    return port_num;
+};
+
+// accessors for AP_AHRS
+bool AP_ExternalAHRS::healthy(void) const
+{
+    uint32_t now = AP_HAL::millis();
+    return (now - last_pkt1_ms < 40 && now - last_pkt2_ms < 500);
+}
+
+bool AP_ExternalAHRS::initialised(void)
+{
+    return last_pkt1_ms != 0 && last_pkt2_ms != 0;
+}
+
+bool AP_ExternalAHRS::get_quaternion(Quaternion &quat)
+{
+    if (!last_pkt1) {
+        return false;
+    }
+    WITH_SEMAPHORE(sem);
+    const struct VN_packet1 &pkt1 = *last_pkt1;
+    quat(pkt1.quaternion[3], pkt1.quaternion[0], pkt1.quaternion[1], pkt1.quaternion[2]);
+    return true;
+}
+
+bool AP_ExternalAHRS::get_origin(Location &loc)
+{
+    WITH_SEMAPHORE(sem);
+    loc = origin;
+    return origin_set;
+}
+
+bool AP_ExternalAHRS::get_location(Location &loc)
+{
+    if (!last_pkt2) {
+        return false;
+    }
+    const struct VN_packet2 &pkt2 = *last_pkt2;
+    WITH_SEMAPHORE(sem);
+    loc = Location();
+    loc.lat = pkt2.GPS1posLLA[0] * 1.0e7;
+    loc.lng = pkt2.GPS1posLLA[1] * 1.0e7;
+    loc.alt = pkt2.GPS1posLLA[2] * 1.0e2;
```

Comment:
We're going to call `get_location` a lot.  This looks like a lot of stuff to be doing all the time, perhaps we should transform when we get the packet.

Also,
```suggestion
    loc = Location{
        pkt2.GPS1posLLA[0] * 1.0e7,
        pkt2.GPS1posLLA[1] * 1.0e7,
        pkt2.GPS1posLLA[2] * 1.0e2,
        Location::AltFrame::ABSOLUTE
    };
```

Reply:
that adds an unnecessary dependency on the order of the elements in Location class